Automattic/sensei
# Fix lesson preview indicator

Closes #2137 

This change makes the Preview indicator for lessons consistent across Sensei.

Since there several different cases to address for this, I have provided a bit of an explanation below of each of the cases (after the testing instructions). Hopefully that will help with code review.

## Testing Instructions

- Create a course with a few lessons. Ensure that some lessons are in a module, and some are not. Check the "Allow this lesson to be viewed without purchase/login" checkbox in the "Preview" metabox for a few lessons: at least one in a module, and at least one that is not in a module.

- There are many cases to test. Please consider the following:

### Pages

The preview indicator should show up on the following pages:

- The Course page (`/course/<course-slug>`). The indicator should show up beside each lesson title. Please ensure that it is there for lessons with and without modules (the code is different for the two).

- The Lesson page (`/lesson/<lesson-slug>`). The indicator should show up beside the title at the top of the page. Please test with a preview lesson and a non-preview lesson. For the preview lesson, the pink notice should not appear (as it was replaced by the new preview indicator) but for the non-preview lesson the pink notice should still be there.

- The Lesson Tag page (`/lesson-tag/<tag-slug>`). The indicator should show up beside the title of the lesson. Ensure that the tag you are testing is set on both preview and non-preview lessons.

### When taking the course

Regardless of whether or not the course is linked to a WooCommerce product, when a user is taking a course, they should not see any preview indicators.

### Logged-out access

Regardless of the state of the Sensei setting "Users must be logged in to view Course and Lesson content", users should always see the preview indicators for courses if they are logged out, or logged in but not taking the course.

### WooCommerce

If the course is linked to a WooCommerce product, the preview indicators should all say "Free Preview" instead of "Preview". However, they should still appear in the same cases as before.

## Code Explanation

The code for the preview indicator was a bit scattered. I've created a helper method to create a consistent HTML tag, which all of the cases should be using now. See the method `sensei_lesson_preview_title_tag` in `includes/class-sensei-frontend.php`.

I've tweaked some CSS to make this indicator look good in all of the various cases. In short, I took the original CSS (which was in `modules-frontend.scss`) and moved it into `frontend/sensei.scss`. I then added a few tweaks for specific cases, building off of that core. See the two SCSS files for that.

The Lesson Page uses a hook on `the_title` to display the indicator using `sensei_lesson_preview_title` in `class-sensei-frontend.php`. That function now uses the helper described above.

On the Course Page, in the section that lists the lessons, there are two cases to consider: lessons with a module, and lessons without a module. For lessons with a module, these already had the appropriate indicator, although it always said "Free Preview" even when WooCommerce was not being used. I fixed this (it uses the new helper now) in `templates/single-course/modules.php`.

For lessons without a module, it uses a filter which called `the_lesson_meta` in `class-sensei-lesson.php`. This was updated to use the new helper. This function is also used on the Lesson Tag page.

The larger change in `class-sensei-lesson.php` is to remove the pink notice for preview lessons. These notices are handled by the `course_signup_link` method. I changed the method to return early for preview lessons, and removed the handling of preview lessons in the three specific cases.

It is worth noting that there is one more piece of code that appears to handle some preview indicator logic: the method [`sensei_course_preview_titles`](https://github.com/Automattic/sensei/blob/e38e907cbcb748352b497867e8e55081435fa7d5/includes/class-sensei-modules.php#L481) within `class-sensei-modules.php`. In my searches, it appeared that this function was only used in a single filter (`sensei_module_lesson_list_title`, added in [`includes/class-sensei-modules.php` line 101](https://github.com/Automattic/sensei/blob/e38e907cbcb748352b497867e8e55081435fa7d5/includes/class-sensei-modules.php#L101)), but this filter is never actually called within Sensei or its extensions. From a Github search, I see that it may be used by some 3rd-party extensions. But its usage is not clear to me, so I left it alone.
